lib/streamlit/commands/execution_control.py

```diff
@@ -54,22 +59,71 @@ def stop() -> NoReturn:  # type: ignore[misc]
         st.empty()
 
 
+def _new_fragment_id_queue(
+    ctx: ScriptRunContext,
+    scope: Literal["app", "fragment"],
+) -> list[str]:
+    if scope == "app":
+        return []
+
+    else:  # scope == "fragment"
+        curr_queue = (
+            ctx.script_requests.fragment_id_queue if ctx.script_requests else []
+        )
+
+        # TODO(vdonato): Figure out if this behavior is reasonable. We'll want to
+        # discuss this with product. The tricky case is that users may run into this
+        # if st.rerun(scope="fragment") is called within a fragment on the first script
+        # run.
+        # Other options include:
+        #   * Rerunning just this fragment even if this was triggered in a full run
+        #     (implementation for this may be tricky)
+        #   * Just rerunning the full script
+        # After more thought, strongly prefer either erroring or doing a full script run
+        # because weird stuff can happen if we try to rerun just a fragment before a
+        # full script run has completed.
+        if not curr_queue:
+            raise StreamlitAPIException(
+                'scope="fragment" can only be specified from `@st.fragment`-decorated '
+                "functions during fragment reruns."
+            )
+
+        assert (
+            new_queue := list(
+                dropwhile(lambda x: x != ctx.current_fragment_id, curr_queue)
+            )
+        ), "Could not find current_fragment_id in fragment_id_queue. This should never happen."
+
+        return new_queue
+
+
+# FIXME(vdonato): e2e tests for this plz.
```

Comment:
would this be covered by modifying the tests (see [comment](https://github.com/streamlit/streamlit/pull/8979/files#r1654428230))?

Reply:
oops -- I staged a removal of this comment but forgot to commit it. The additions to `e2e_playwright/st_rerun.py` were the e2e tests that I was hoping to add here.